### PR TITLE
vala: Add support for cross-compilation.

### LINKED ISSUE #1771: Cross file does not allow passing Vala compiler and args
Which makes sense based on the current logic:

```py
        elif lang == 'vala':
            comp = self.environment.detect_vala_compiler()
            if need_cross_compiler:
                cross_comp = comp  # Vala compiles to platform-independent C
```

But I would argue that it is very useful to be able to pass `-DFOO` and
do `#if FOO` in Vala code, for the few cases where platform-dependent
Vala code is needed.

### PR DESCRIPTION
**Pull request by oleavr**
It is very useful to be able to pass `-DFOO` and do `#if FOO` in Vala code,
for the few cases where platform-dependent Vala code is needed.

Fixes #1771

### CODE REVIEW

File: mesonbuild/environment.py [hunk 1]
```diff
@@ -632,16 +633,22 @@ def detect_cs_compiler(self):
...
+                compiler = [compiler]
+            arg = ['--version']
+            try:
+                p, out = Popen_safe(compiler + arg)[0:2]
+            except OSError as e:
+                popen_exceptions[' '.join(compiler + arg)] = e
+                continue
+            version = search_version(out)
+            if 'Vala' in out:
+                return ValaCompiler(compiler, version, is_cross)
```
Comment (ignatenkobrain) [inline mesonbuild/environment.py]: __[flake8]__
* [F405] 'ValaCompiler' may be undefined, or defined from star imports: .compilers

Comment (oleavr) [inline mesonbuild/environment.py]: I realize this is an automated check, however, the offending line of code was not introduced by my PR, I just happened to bump its indentation. I'm hesitant to introduce unrelated changes in this PR – would you prefer that I also change this?

**Comment by oleavr** [comment]
Note: I realized my commit message style of the previous two PRs were
slightly off from the convention followed by @jpakkane. I've adjusted it
for these two PRs. Please let me know if I got it wrong. 😓

**Comment by arteymix** [comment]
That's excellent! I'm in the process of porting Valum to Windows, so this will be very useful for falling back on `gio-windows-2.0`. I'll test it once it's ready.

**Comment by oleavr** [comment]
(Amended the commit without changing anything seeing as the appveyor failure looks like a fluke.)

**Comment by jpakkane** [comment]
This would require a test of some sort. Maybe seeing that the compiler flags for cross vala compilation go to cross compilations and native flags to native.

**Comment by oleavr** [comment]
@jpakkane Would it be safe to assume that the CI machine(s) that are set up for `cross/ubuntu-armhf.txt` also have `glib-2.0` available when asking` arm-linux-gnueabihf-pkg-config` for it? Or maybe there's a simpler way we can test this?

**Comment by jpakkane** [comment]
Glib 2 is available, but only the native version. I'm not entirely sure how that works in Ubuntu. Do you need to install a separate package to get the cross build pc files?

**Comment by oleavr** [comment]
@jpakkane Yeah we need to get the .pc files so Meson finds them. There isn't a package available for this though. But maybe we can use the host's pkg-config in the cross file and pretend everything's okay, since we're not going to do the actual compilation...

**Comment by jpakkane** [comment]
That is flaky. This also needs an entry in the release notes as it is new functionality.

**Comment by jpakkane** [comment]
Ping.

For the test you don't need a full project. Have a unit test that runs just the configure step and then inspect that the build compilation data base has the requested arguments.

**Comment by jpakkane** [comment]
:crossed_swords: conflicts.

**Comment by astavale** [comment]
What's the status of this? I'd like to have VALAC environment variable working with Meson for 0.42 if at all possible. This PR stopped #1851. For getting VALAC working with 0.42 I see two options:

 1. This PR is rebased, if the author has time, although I'm not sure how easy the unit tests are
 2. Commit #1851 for 0.42 instead and so give the author of this PR more time to rebase and write the unit tests

**Comment by oleavr** [comment]
Sorry for the delay here, I'm juggling too many things these days. I won't be able to get to this in a while, so unless someone volunteers to finish this I'd vote for landing #1851 first.

**Comment by astavale** [comment]
@jpakkane if you need updates to the release notes and/or vala documentation for #1851 i can do that straight away

**Comment by jpakkane** [comment]
#1851 has landed. What shall we do with this?

**Comment by jpakkane** [comment]
ping

**Comment by jpakkane** [comment]
Closing due to inactivity to reduce clutter on the MR tracker. Feel free to reopen if you have a fixed version. Thanks.
